db/sql/0_msar.sql

```diff
@@ -902,29 +1032,49 @@ All fields are optional, and a null value as input returns 'text'
 */
 SELECT COALESCE(
   -- First choice is the type specified by numeric IDs, since they're most reliable.
-  format_type(typ.id, typ.modifier),
+  format_type((typ_jsonb ->> 'id')::integer, (typ_jsonb ->> 'modifier')::integer),
   -- Second choice is the type specified by string IDs.
-  COALESCE(
-    msar.get_fully_qualified_object_name(typ.schema, typ.name),
-    typ.name,
-    'text'  -- We fall back to 'text' when input is null or empty.
-  )::regtype::text || COALESCE(  -- This section builds the type options blob.
-    '(' || topts.length || ')',
-    ' ' || topts.fields || ' (' || topts.precision || ')',
-    ' ' || topts.fields,
-    '(' || topts.precision || ', ' || topts.scale || ')',
-    '(' || topts.precision || ')',
-    ''
-  ) || CASE WHEN topts.array_ THEN '[]' ELSE '' END
+  __msar.get_formatted_base_type(
+    COALESCE(
+      msar.get_fully_qualified_object_name(typ_jsonb ->> 'schema', typ_jsonb ->> 'name'),
+      typ_jsonb ->> 'name',
+      'text'  -- We fall back to 'text' when input is null or empty.
+    ),
+    typ_jsonb -> 'options'
+  ) || CASE WHEN (typ_jsonb -> 'options' ->> 'array')::boolean THEN '[]' ELSE '' END
 )
-FROM
-  jsonb_to_record(typ_jsonb)
-    AS typ(id oid, schema text, name text, modifier integer, options jsonb),
-  jsonb_to_record(typ_jsonb -> 'options')
-    AS topts(length integer, precision integer, scale integer, fields text, array_ boolean);
 $$ LANGUAGE SQL;
 
 
+CREATE OR REPLACE FUNCTION
+msar.build_type_text_complete(typ_jsonb jsonb, old_type text) RETURNS text AS $$/*
+Build the text name of a type, using the old type as a base if only options are given.
+
+The main use for this is to allow for altering only the options of the type of a column.
+
+Args:
+  typ_jsonb: This is a jsonb denoting the new type.
+  old_type: This is the old type name, with no options.
+
+The typ_jsonb should be in the form:
+{
+  "name": <str> (optional),
+  "options": <obj> (optional)
+}
+
+*/
+SELECT msar.build_type_text(
+  jsonb_strip_nulls(
+    jsonb_build_object(
+      'name', COALESCE(typ_jsonb ->> 'name', old_type),
```

Comment:
I'm surprised that the `typ_jsonb` struct doesn't consider the type name a required field. When does that offer advantages? I see how that makes handling it more fiddly, however. You mention the use case where you want to only change the options of a type, but in such a case the type name/id would be easily accessible, right? What's the win here?

By the way, loving the documentation.

Reply:
It's accessible in the database layer, but this JSON was constructed in the Python layer. If we have to look up the type of a column during the call where it's generated, it's a round trip to the DB. Plus, I find these lookups to be cleaner and generally _less_ fiddly here in the DB than they were in python.

Another clean solution would be for the front end to add the type name when it originally calls the API, I suppose. There may be other problems with that; not sure.

Reply:
I think you're right. There is a smell here:
- It's a bit inconsistent when we consider type options to be attached to the 'whole' type.
- E.g., if we change from `CHARACTER(256)` to `CHARACTER VARYING(256)`, the frontend has to submit the type option length=256 again.